R/forward_selection.R

```diff
@@ -0,0 +1,121 @@
+#' Select features based on feature selection
+#'
+#' @description Select features using forward selection algorithm.
+#'   It starts as an empty model, and add the variable with the
+#'   highest improvement in the accuracy of the model. The process
+#'   is iteratively repeated and it stops when the remaining variables
+#'   doesn't improve the accuracy of the model.
+#'
+#' @param scorer  A custom user-supplied function that accepts X and y
+#'   (as defined below) as input and returns the error of the datasets.
+#' @param X tibble. training dataset
+#' @param y tibble. test dataset
+#' @param min_features double. number of minimum features to select
+#' @param max_features double. number of maximum features to select
+#'
+#' @return vector. The indexes of selected features
+#' @export
+#' @examples
+#' forward_select(
+#'   train_data, test_data, max_features=5
+#' )
+
+forward_select <- function(scorer, X, y, min_features=1, max_features=10) {
+  # Tests
+  # 'scorer' must be a function
+  if (class(scorer) != "function") {
+    stop("Expected a function for `scorer`." )
+  }
+
+  # X and y must be a data.frame or something compatible
+  if (!any(class(X) == "data.frame")) {
+    stop("Expected a 'data.frame' object for `data`.")
+  }
+
+  if (sum(!(dim(X))==1) != 2){
+    stop("X must be a 2-d array")
+  }
+
+  if (!any(class(y) == "data.frame")) {
+    stop("Expected a 'data.frame' object for `data`.")
+  }
+
+  if (sum(!(dim(y))==1) != 1){
+    stop("y must be a 1-d array")
+  }
+
+  # Looks for X and y have the same number of samples
+  if (dim(X)[1] != dim(y)[1]){
+    stop("X and y have inconsistent numbers of samples. X:", dim(X)[1], ", y:", dim(y)[1])
+  }
+
+  # max_features should be greather or equal to min_features
+  if (min_features > max_features){
+    stop("max_features should be greater or equal to min_features.")
+  }
+
+  # min_features should be positive, we want to select at least one feature
+  if (min_features < 1){
+    stop("min_features should be a positive number.")
+  }
+
+  # Initialize parameters
+  # Obtain initial array of randomly selected features
+  ftr_no_selection <- c(1:length(X))
+  ftr_selection <- c()
+  ftr_running <- c()
+  scores_fn <- c()#Inf
+  X_new <- c()
+  best_scores_iter <- Inf
+  best_scores_all <- c()
+  flag_keep_running = TRUE
```

Comment:
Why do you need two flags? It is quite confusing. Doesn't `flag_keep_running = TRUE` mean `flag_stop_running = FALSE` inherently? When is it not the case?